simulator-core/src/main/java/com/evbox/everon/ocpp/simulator/station/handlers/ocpp/ChangeAvailabilityRequestHandler.java

```diff
@@ -0,0 +1,112 @@
+import com.evbox.everon.ocpp.v20.message.station.ChangeAvailabilityResponse;
+import lombok.extern.slf4j.Slf4j;
+
+import static com.evbox.everon.ocpp.v20.message.station.ChangeAvailabilityResponse.Status.*;
+
+/**
+ * Handler for {@link ChangeAvailabilityRequest} request.
+ */
+@Slf4j
+public class ChangeAvailabilityRequestHandler implements OcppRequestHandler<ChangeAvailabilityRequest> {
+
+    private static final int DEFAULT_NUMBER_OF_ATTEMPTS = 30;
+    private static final int DELAY_BETWEEN_ATTEMPTS = 1_000;
+
+    private final StationState stationState;
+    private final StationMessageSender stationMessageSender;
+    private final int attemptsInMillis;
+    private final int delayBetweenAttemptsInMillis;
+
+    private final ChangeAvailabilityFuture changeAvailabilityFuture = new ChangeAvailabilityFuture();
+
+    /**
+     * Create an instance.
+     *
+     * @param stationState         {@link StationState}
+     * @param stationMessageSender {@link StationMessageSender}
+     */
+    public ChangeAvailabilityRequestHandler(StationState stationState, StationMessageSender stationMessageSender) {
+        this(stationState, stationMessageSender, DEFAULT_NUMBER_OF_ATTEMPTS, DELAY_BETWEEN_ATTEMPTS);
+    }
+
+    /**
+     * Create an instance.
+     *
+     * @param stationState                 {@link StationState}
+     * @param stationMessageSender         {@link StationMessageSender}
+     * @param attemptsInMillis             number of attempts to try to change the state before giving up
+     * @param delayBetweenAttemptsInMillis delay between attempts
+     */
+    public ChangeAvailabilityRequestHandler(StationState stationState, StationMessageSender stationMessageSender,
+                                            int attemptsInMillis, int delayBetweenAttemptsInMillis) {
+        this.stationState = stationState;
+        this.stationMessageSender = stationMessageSender;
+        this.attemptsInMillis = attemptsInMillis;
+        this.delayBetweenAttemptsInMillis = delayBetweenAttemptsInMillis;
+    }
+
+    /**
+     * Handle {@link ChangeAvailabilityRequest} request.
+     * It has 4 scenarios:
+     * 1. Send response with ACCEPTED status when EVSE state is the same as requested.
+     * 2. Change EVSE state to the requested state when they do not match.
+     * In addition send response with ACCEPTED status and StatusNotification request for every EVSE Connector.
+     * 3. When a transaction is in progress.
+     * Send response with SCHEDULED status and schedule a task to change the EVSE state when transaction is finished.
+     * 4. Send response with REJECTED status if exception occurs.
+     *
+     * @param callId  identity of the message
+     * @param request incoming request from the server
+     */
+    @Override
+    public void handle(String callId, ChangeAvailabilityRequest request) {
+        try {
+            Evse evse = stationState.findEvse(request.getEvseId());
```

Comment:
I would refactor findEvse to return Optional instead of throwing exception and rethrowing it further

Reply:
I agree. I started refactoring StationState class but in another MR.